vcf.c

```diff
@@ -2626,20 +2637,91 @@ int bcf_enc_vchar(kstring_t *s, int l, const char *a)
     return 0; // FIXME: check for errs in this function
 }
 
+// Special case of n==1 as it also occurs quite often in FORMAT data.
+// This version is also small enough to get inlined.
+static inline int bcf_fmt_array1(kstring_t *s, int type, void *data) {
+    uint32_t e = 0;
+    uint8_t *p = (uint8_t *)data;
+    int32_t v;
+
+    // helps gcc more than clang here. In billions of cycles:
+    //          bcf_fmt_array1  bcf_fmt_array
+    // gcc7:    23.2            24.3
+    // gcc13:   21.6            23.0
+    // clang13: 27.1            27.8
+    switch (type) {
+    case BCF_BT_CHAR:
+        e |= kputc_(*p == bcf_str_missing ? '.' : *p, s) < 0;
+        break;
+
+    case BCF_BT_INT8:
+        if (*(int8_t *)p != bcf_int8_vector_end) {
+            e |= ((*(int8_t *)p == bcf_int8_missing)
+                  ? kputc_('.', s)
+                  : kputw(*(int8_t *)p, s)) < 0;
+        }
+        break;
+    case BCF_BT_INT16:
+        v = le_to_i16(p);
+        if (v != bcf_int16_vector_end) {
+            e |= (v == bcf_int16_missing
+                  ? kputc_('.', s)
+                  : kputw(v, s)) < 0;
+        }
+        break;
+
+    case BCF_BT_INT32:
+        v = le_to_i32(p);
+        if (v != bcf_int32_vector_end) {
+            e |= (v == bcf_int32_missing
+                  ? kputc_('.', s)
+                  : kputw(v, s)) < 0;
+        }
+        break;
+
+    case BCF_BT_FLOAT:
+        v = le_to_u32(p);
+        if (v != bcf_float_vector_end) {
+            e |= (v == bcf_float_missing
+                  ? kputc_('.', s)
+                  : kputd(le_to_float(p), s)) < 0;
+        }
+        break;
+
+    default:
+        hts_log_error("Unexpected type %d", type);
+        return -1;
+    }
+
+    return e == 0 ? 0 : -1;
+}
+
 int bcf_fmt_array(kstring_t *s, int n, int type, void *data)
 {
     int j = 0;
     uint32_t e = 0;
     if (n == 0) {
-        return kputc('.', s) >= 0 ? 0 : -1;
+        return kputc_('.', s) >= 0 ? 0 : -1;
```

Comment:
As this function is part of the API, this should remain as `kputc()` so the caller gets a properly terminated string.